// sveltekit-sitemap/src/plugin.ts
import fs from "fs";
import type { ViteDevServer } from "vite";
import { Sitemap, SitemapPluginParams } from "./types";

// function to recursively traverse a directory and return if some of the files are named +page.svelte
const hasPageInside = (path: string): boolean => {
  const files = fs.readdirSync(path);
  if (files.some((file) => file === "+page.svelte")) return true;
  return files.some((file: string) => {
    const newPath = path + "/" + file;
    if (fs.statSync(newPath).isDirectory()) {
      return hasPageInside(newPath);
    } else {
      return false;
    }
  });
};

const getRoutes = (dir: string): Sitemap => {
  let routes: Sitemap = {};
  const traverseRoutes = (path: string) => {
    const isDirectory = fs.statSync(path).isDirectory();
    const isPageFolder = isDirectory && hasPageInside(path);
    if (isDirectory && isPageFolder) {
      fs.readdirSync(path).forEach((file) => traverseRoutes(path + "/" + file));
    }

    const id = path.replace(dir, "").replace("/+page.svelte", "");

    const dirBase = path.replace("/+page.svelte", "");

    const isFolder =
      fs.statSync(dirBase).isDirectory() &&
      fs.readdirSync(path.replace("/+page.svelte", "")).some((p) => {
        return fs.statSync(dirBase + "/" + p).isDirectory();
      });
    if (!path.endsWith("+page.svelte") && !isFolder) return;

    Object.assign(routes, { [id || "/"]: isFolder });
  };
  fs.readdirSync(dir).forEach((file) => traverseRoutes(dir + "/" + file));

  return routes;
};

export const sitemapPlugin = ({
  routesDir = "./src/routes",
  sitemapFile = "./src/sitemap.ts"
}: SitemapPluginParams = {}) => {
  function updateSitemap() {
    fs.writeFileSync(
      sitemapFile,
      `import type { RO_Sitemap } from 'sveltekit-sitemap';

export const sitemap = (<const>${JSON.stringify(getRoutes(routesDir), null, 3).replace(
        /\uFFFF/g,
        '\\"'
      )}) satisfies RO_Sitemap

export type Sitemap = typeof sitemap
`
    );
  }
  updateSitemap();

  return {
    name: "sveltekit-sitemap",
    configureServer(server: ViteDevServer) {
      server.watcher
        .add([routesDir])
        .on("add", updateSitemap)
        .on("unlink", updateSitemap)
        .on("unlinkDir", updateSitemap);
    }
  };
};
